apps/game_backend/lib/game_backend/matches.ex

```diff
@@ -2,21 +2,48 @@ defmodule GameBackend.Matches do
   @moduledoc """
   Matches
   """
+  alias GameBackend.Users
+  alias GameBackend.Users.Currencies
   alias Ecto.Multi
   alias GameBackend.Repo
   alias GameBackend.Matches.ArenaMatchResult
 
-  def create_arena_match_results(match_id, results) do
-    transaction = Multi.new()
+  def create_arena_match_results(results) do
+    {:ok, conrrency_config_json} =
+      Application.app_dir(:game_backend, "priv/currencies_rules.json")
+      |> File.read()
 
-    transaction =
-      Enum.reduce(results, transaction, fn result, transaction_acc ->
-        attrs = Map.put(result, "match_id", match_id)
-        changeset = ArenaMatchResult.changeset(%ArenaMatchResult{}, attrs)
-        transaction_acc = Multi.insert(transaction_acc, {:insert, attrs["user_id"]}, changeset)
-        transaction_acc
-      end)
+    currency_config =
+      Jason.decode!(conrrency_config_json)
 
-    Repo.transaction(transaction)
+    Enum.reduce(results, Multi.new(), fn result, transaction_acc ->
+      changeset = ArenaMatchResult.changeset(%ArenaMatchResult{}, result)
+      {:ok, google_user} = Users.get_google_user(result["user_id"])
+
+      amount_of_trophies = Currencies.get_amount_of_currency_by_name(google_user.user.id, "Trophies")
+
+      amount =
+        get_amount_of_trophies_to_modify(amount_of_trophies, result["position"], currency_config)
+
+      Multi.insert(transaction_acc, {:insert, result["user_id"]}, changeset)
+      |> Multi.run(
+        {:add_trophies_to, result["user_id"]},
+        fn _, _ ->
+          Currencies.add_currency_by_name!(google_user.user.id, "Trophies", amount)
+        end
+      )
+    end)
+    |> Repo.transaction()
+  end
+
+  def get_amount_of_trophies_to_modify(current_trophies, position, currencies_config) do
```

Comment:
Kinda tied to my comment on the pre-processing of the configuration, if we did that and have it in a easier format like that then we can simplify some of the logic here and have something like

```elixir
  def get_amount_of_trophies_to_modify(current_trophies, position) do
    trophies_config = Application.get_env(:arena, GameBackend.Currencies)[:trophies]
    distribution = Enum.find(trophies_config[:distributions], trophies_config[:infinity], fn max_amount -> max_amount > current_trophies end)
    get_in(trophies_config, [distribution, position])
  end
```

Reply:
this was changed thanks to the previous comment!